### DISCUSSION TITLE
Channel Impulse Response moving scenario

### DISCUSSION THREAD

**Discussion by FelixSchmidmeir** [NVlabs/sionna#413] [General]
Hi there,

I am currently trying to create a moving Channel based Key Generation scenario (with CIR) in Sionna. 

As Sionna, as far as I know, does not natively support moving scenarios I am not sure if the way I am doing it is correct.

At the moment I have two parties Alice and Bob, which are represented by a transmitter and a receiver each. The plan is to add adversaries later on to compare measurements. While Alice is stationary Bob is supposed to walk on a line from a start to an end position, which is represented in the picture below.

![Screenshot from 2024-04-15 14-19-33](https://github.com/NVlabs/sionna/assets/38666033/66fa361e-0f69-4fac-b06c-bcd056124d8b)

To simulate the movement I calculate the positions of Bob on the line and run a simulation for each of these positions. Parts of the simulation results are later combined to a final result the following way

result = part_1_sim_1 + part_2_sim_2 + part_3_sim_3 …

The expected result would be, that the CIRs from alicetx → bobrx and bobtx → alicerx are equal or very similar, while for example the CIR alicetx → eve_1_rx is completely different. Furthermore the result values should be an adequate representation for a moving scenario.

So my question would be if this is an adequate simulation for a moving scenario and if not if there is a better way? Or if this is even possible using Sionna?

Additionally I came across the function apply_doppler(), which I think would be needed, but I am not sure how to properly apply it.

At the moment I am generating a position for Bob for each meter of the line. So I would use the number of positions, which is equal to the length of the line, as the number of time steps? How many positions should I compute for Bob?

With the number of time steps I could also calculate the tx and rx velocities. The problem would only be that they represent the velocities of all tx and rx and in my scenario only Bob is moving while Alice is stationary.

Thank you very much for any help in advance.

```py
from sionna.rt import load_scene, Transmitter, Receiver, PlanarArray, Camera
from sionna.channel.utils import subcarrier_frequencies
from sionna.channel.utils import cir_to_ofdm_channel
import sionna
import tensorflow as tf
import math

# calculates a line between the starting and end position of b
# returns a list of positions on the line
def get_positions(start_pos, end_pos):
    x = end_pos[0] - start_pos[0]
    y = end_pos[1] - start_pos[1]
    z = end_pos[2] - start_pos[2]

    length = math.ceil(math.sqrt(x ** 2 + y ** 2 + z ** 2))

    x /= length
    y /= length
    z /= length

    positions = []
    for i in range(0, length):
        x_pos = start_pos[0] + i * x
        y_pos = start_pos[1] + i * y
        z_pos = start_pos[2] + i * z
        positions.append([x_pos, y_pos, z_pos])
    return positions


def initialize_scene(pos_a, pos_b):
    simulation_scene = load_scene(sionna.rt.scene.etoile)
    simulation_scene.tx_array = PlanarArray(num_rows=1,
                                            num_cols=1,
                                            vertical_spacing=0.5,
                                            horizontal_spacing=0.5,
                                            pattern="tr38901",
                                            polarization="V")

    # Configure antenna array for all receivers
    simulation_scene.rx_array = PlanarArray(num_rows=1,
                                            num_cols=1,
                                            vertical_spacing=0.5,
                                            horizontal_spacing=0.5,
                                            pattern="tr38901",
                                            polarization="V")

    # Create transmitters
    alicetx = Transmitter(name="aliceTX",
                          position=pos_a,
                          orientation=[0, 0, 0])

    simulation_scene.add(alicetx)

    bobtx = Transmitter(name="bobTX",
                        position=pos_b,
                        orientation=[0, 0, 0])

    simulation_scene.add(bobtx)

    # Create a receivers
    alicerx = Receiver(name="aliceRX",
                       position=pos_a,
                       orientation=[0, 0, 0])

    simulation_scene.add(alicerx)

    bobrx = Receiver(name="bobRX",
                     position=pos_b,
                     orientation=[0, 0, 0])

    simulation_scene.add(bobrx)

    ## RX points towards TX
    bobrx.look_at(alicetx)
    alicerx.look_at(bobtx)

    ## TX points towards RX
    bobtx.look_at(alicerx)
    alicetx.look_at(bobrx)

    frequency = 2e9
    simulation_scene.frequency = frequency
    simulation_scene.synthetic_array = True

    current_paths = simulation_scene.compute_paths()

    # TODO apply doppler effect?
    #current_paths.apply_doppler(frequency, ..., ..., ...)

    a, tau = current_paths.cir()
    frequencies = subcarrier_frequencies(fft_size, 15e3)
    current_cir = cir_to_ofdm_channel(frequencies, a, tau)
    current_cir = tf.reshape(current_cir, (4, 1000))

    # return cir for bobtx -> alicerx and alicetx -> bobrx
    return current_cir[1], current_cir[2]

position_a = [0, 0, 1.5]
# starting and end position of bob
start_b = [-20, -20, 1.5]
end_b = [-20, 20, 1.5]

fft_size = 1000

# calculate positions for bob on the line between start and end
positions_b = get_positions(start_b, end_b)

# run a simulation for each position of bob
results = []
for position_b in positions_b:
    results.append(initialize_scene(position_a, position_b))

res_length = len(results)
# calculate how many measurements to use per simulation
nr_measurements = fft_size // res_length

# combine parts of simulations to final results
result_a = []
result_b = []
for index in range(0, res_length):
    start_index = nr_measurements * index
    result_a.extend(results[index][0][start_index:start_index + nr_measurements])
    result_b.extend(results[index][1][start_index:start_index + nr_measurements])

```

**Comment by FelixSchmidmeir** [comment]
Hi there, 

I am still working on this Issue and currently I am trying to figure out, how to use apply_doppler() correctly.

### num_time_steps
One of the parameters of apply_doppler is num_time_steps. Num_time_steps is also an output of the function cir_to_ofdm_channel. As the output for num_time_steps in my scenarios is 1, I am assuming 1 is also the correct input for apply_doppler in my case?

### tx_velocities and rx_velocities
With these parameters the velocities of all receivers and all transmitters are defined. In my scenario the transmitters and receivers of alice are stationary while the ones of bob are moving. Therefore not all receivers and transmitters have the same velocities.

Therefore my question would be if there is a way to define different velocities for the receivers and transmitters of alice and bob?

Furthermore I am not sure how to calculate the right velocities. If Bob travels for example from [0, 0, 1.5] to [15, 0, 1.5], would this mean as I only have one timestep, that the velocity of his transmitters and receivers is [15, 0, 0]?

@jhoydis do you maybe have any idea how to go on with this?

Thank you very much in advance and any help would be appreciated

**Reply by jhoydis** [reply]
As per the [API Documentation](https://nvlabs.github.io/sionna/api/rt.html#sionna.rt.Paths.apply_doppler), the tensors of velocities can define a different velocity for each transmitter and receiver.

**Comment by jhoydis** [comment]
Hi @FelixSchmidmeir,

Moving a  receiver (or transmitter) along a trajectory is the right way of tackling this. 
If you would like to get one CIR every T [s] and assume that the receiver moves at a speed of v [m/s], you would need to create samples along the trajectory spaced v*T [m] apart.

The `apply_doppler` function creates also a sequence of CIR time samples assuming a certain speed of transmitter and/or receiver. However, it works without tracing of new paths, making it much faster. The caveat is that it is only accurate over very short time horizons.
